### ISSUE TITLE
Add callback on image / file saved

### ISSUE THREAD

**Issue by shcheklein** [ultralytics/ultralytics#1909]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar feature requests.

### Description

We need to see when trainer, validator, etc create images (samples, metric plots, etc) to be able to capture them in the loggers. I'm creating a logger (as well as some other things, like [`devcontainer` with GPU support](https://github.com/ultralytics/ultralytics/compare/main...shcheklein:ultralytics:main#diff-8c75d30eaaa096649631dde76e321d13a6849c64137f3f7f777429ef9fd6b9e2) that I plan also to contribute) for DVC [here](https://github.com/ultralytics/ultralytics/compare/main...shcheklein:ultralytics:main#diff-052c6262efa033fb56c6ac6241d151d7eb373ee01daa36cf70077f332078a216). I think we (and other loggers) can benefit a lot from this.

### Use case

Pretty much similar to what Comet logger already does in an ad-hoc way, e.g.:

```python
def _log_plots(experiment, trainer):
    plot_filenames = [trainer.save_dir / f'{plots}.png' for plots in EVALUATION_PLOT_NAMES]
    _log_images(experiment, plot_filenames, None)

    label_plot_filenames = [trainer.save_dir / f'{labels}.jpg' for labels in LABEL_PLOT_NAMES]
    _log_images(experiment, label_plot_filenames, None)
```

by hardcoding `EVALUATION_PLOT_NAMES`, etc. It's doable and a possible workaround, but it's fragile (if code changes in the core, it clearly breaks the logger). Also, it would be sad if we have to duplicate the same logic everywhere.

### Additional

Suggestion is to add a new type of callback, e.g. `on_train_image_save` or `on_train_plot_save`. An alternative is to have some instance dict where we could register all visualizations (e.g. trainer.visualizations? or `trainer.plots`)- downside can be that we don't know which image is per epoch, which one is at the end, etc.

### Are you willing to submit a PR?

- [X] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
@shcheklein, thank you for bringing this feature request to our attention! We appreciate your interest in contributing to YOLOv8, and we encourage you to submit a pull request with your proposed code changes. 

Regarding your use case, we agree that it would be more robust and maintainable to have a standardized way of logging visualizations rather than hard-coding them in ad-hoc ways in specific loggers. The idea of adding new callbacks like `on_train_image_save` or `on_train_plot_save` seems reasonable, and we will consider this for a future update to YOLOv8.

In the meantime, please feel free to submit your proposed changes, and we will review them carefully. Let us know if you have any further questions or concerns!

**Comment by glenn-jocher** [comment]
@shcheklein on reviewing this I'm not sure this makes sense, because all current callbacks are based on a specific event which we know the location of, i.e. we know where the line for batch end, epoch end, etc. exists, whereas images actions like saving or generating are not determined by a location in the code and are arbitrary. 

Right now there is a fixed set of images generated which we know the location of, but there is no guarantee that this won't expand in the future. 

If we have added maintenance of placing a callback after every image that may be added in the future then we incur risk of detaching future image generation from callback addition. 

I'd say you may want to look into monkey-patching known image saving functions like cv2.imwrite instead to obtain your desired result. This should resolve all of the concerns mentioned above.

@AyushExel what do you think of this?

**Comment by AyushExel** [comment]
@glenn-jocher I think the hardcoding of plots in the Global variable might be a good idea, like trainer.plots, validator.plots etc, and then use the existing callbacks for logging them.
@shcheklein Yes please make a PR with the proposed changes, we'd love to support more integrations. The initial code diff on your branch link looks good.

**Comment by shcheklein** [comment]
Thanks folks, I'm moving with my fork first to get more sense about this since I also want to be a bit more careful and not complicate things w/o a good reason. Monkey patching might work of course, the difference with the callback is that we can and we need a bit more context for loggers (is it a train stage? is it a sample or a plot? etc - we can potentially pass this additional info into a callback call - that would be very handy).

> Global variable might be a good idea, like `trainer.plots`, `validator.plots` etc, and then use the existing callbacks for logging them.

You mean a variable in some base class, right? So that all loggers could reuse them?

---------

If we go with an additional callback, we can and probably should make a wrapper in some utils - like `save_image` that would do things like: trigger a callback, write a file, etc - so that we can consolidate all the logic in one place.

**Comment by glenn-jocher** [comment]
@shcheklein hey buddy. Yes after talking to @AyushExel I think it would make sense to consolidate all images and plots into a centralized place like `trainer.plots` for later consumption in logger callbacks. 

There is a lot of duplication at the moment between different loggers attempting to determine on their own what images are available from a hard-coded list, which is suboptimal.

**Comment by shcheklein** [comment]
Apologies for the delay here. As things are progressing on the DVC logger I'm getting more sense what is needed and see different places where plots are being produced, their semantics, etc.

How about this approach https://github.com/ultralytics/ultralytics/pull/2461 ? I can extend it to deal with validator, etc. Seems  more or less manageable and flexible enough to handle validator plots that are scattered deep inside Metrics utils, etc.

We can extend the dict later with data for the plots so that loggers even can do some custom stuff with actual plots data vs just file images (it would be handy for DVC logger for example since it potentially can plot multiple runs on the same figure to compare them).

Let me know if that works in general, I'll expand it to handle other plots.

**Comment by glenn-jocher** [comment]
@shcheklein thank you for your message and proposed approach for handling plots in YOLOv8. We appreciate your effort in improving our library.

Based on your proposed solution in the PR that you linked, it seems like a manageable and flexible approach for handling plots. It's good to hear that you're extending the solution to handle plots in the Validator and Metrics utils, which will make sure that everything is thoroughly covered.

The idea of extending the dictionary later with data for the plots is a great one, as it would allow loggers to access actual plot data and perform custom operations, in addition to working with generated file images.

Overall, we think your approach is a step in the right direction and we're excited to see it implemented. Thank you again for your continued contributions and please let us know if you have any further questions or concerns.

**Comment by shcheklein** [comment]
Should be fixed by #2461
